icu4c/source/i18n/unicode/messageformat2.h

```diff
@@ -0,0 +1,138 @@
+// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#ifndef U_HIDE_DEPRECATED_API
+
+#ifndef MESSAGEFORMAT2_H
+#define MESSAGEFORMAT2_H
+
+#if U_SHOW_CPLUSPLUS_API
+
+/**
+ * \file
+ * \brief C++ API: Formats messages in a language-neutral way using the draft MessageFormat 2.0.
+ */
+
+#if !UCONFIG_NO_FORMATTING
+
+#include "unicode/format.h"
+#include "unicode/parseerr.h"
+#include "unicode/unistr.h"
+#include "unicode/utypes.h"
+
+U_NAMESPACE_BEGIN
+
+/**
+ * <p>MessageFormat2 is a Technical Preview API implementing MessageFormat 2.0.
+ * Since it is not final, documentation has not yet been added everywhere.
+ *
+ * <p>See <a target="github" href="https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md">the
+ * description of the syntax with examples and use cases</a> and the corresponding
+ * <a target="github" href="https://github.com/unicode-org/message-format-wg/blob/main/spec/message.abnf">ABNF</a> grammar.</p>
+ *
+ * @internal ICU 74.0 technology preview
+ * @deprecated This API is for technology preview only.
+ */
+
+class U_I18N_API MessageFormat2 : public Format {
+  public:
+    /**
+     * Constructs a new MessageFormat2 using the given pattern and the
+     * default locale.
+     *
+     * @param pattern   Pattern used to construct object.
+     * @param parseError Struct to receive information on the position
+     *                   of an error within the pattern.
+     * @param status    Input/output error code.  If the
+     *                  pattern cannot be parsed, set to failure code.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+    MessageFormat2(const UnicodeString &pattern, UParseError &parseError, UErrorCode &status);
+
+    /**
+     * Destructor.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+    virtual ~MessageFormat2();
+
+    // Not yet implemented
+    virtual bool operator==(const Format &other) const override;
+    // Not yet implemented
+    virtual bool operator!=(const Format &other) const;
+
+    // Not yet implemented
+    virtual UnicodeString &format(const Formattable &obj, UnicodeString &appendTo, FieldPosition &pos,
+                                  UErrorCode &status) const override;
+
+    // Not yet implemented
+    virtual void parseObject(const UnicodeString &source, Formattable &result,
+                             ParsePosition &pos) const override;
+
+    /**
+     * Clones this Format object polymorphically.  The caller owns the
+     * result and should delete it when done.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+
+    virtual MessageFormat2 *clone() const override;
```

Comment:
I pinged you on Slack to ask more about the design docs.

Regarding `clone()`, this is because I was assuming that `MessageFormat2` should inherit from `Format`, which has a pure virtual `clone()` method. I see that the `MessageFormatter` class in the ICU4J impl doesn't inherit from any other classes. (I'm also not sure what the Java equivalent of the `Format` class is, or if there is one.) If there's no need to inherit from `Format`, then I can remove `clone()`.